spark/src/main/scala/org/apache/spark/sql/delta/OptimisticTransaction.scala

```diff
@@ -523,7 +523,20 @@ trait OptimisticTransactionImpl extends TransactionalWrite
 
       val newProtocolForLatestMetadata =
         Protocol(readerVersionAsTableProp, writerVersionAsTableProp)
-      val proposedNewProtocol = protocolBeforeUpdate.merge(newProtocolForLatestMetadata)
+
+      // When the protocol versions are explicitly set on table features protocol we may
+      // downgrade to legacy protocol versions. Legacy protocol versions can only be
+      // used if a table supports *exactly* the set of features in that legacy protocol
+      // version, with no "gaps". By merging in the protocol features from a particular
+      // protocol version, we may end up with such a "gap-free" protocol. E.g. if a table
+      // has only table feature "checkConstraints" (added by writer protocol version 3)
+      // but not "invariants" and "appendOnly", then setting the minWriterVersion to
+      // 2 or 3 will add "invariants" and "appendOnly", filling in the gaps for writer
+      // protocol version 3, and then we can downgrade to version 3.
+      val proposedNewProtocol = protocolBeforeUpdate
+        .merge(newProtocolForLatestMetadata)
```

Comment:
It seems to me that this line and the next line should do the same thing. Looking at `merge`, it will add all implicitly supported features in any of the merged protocols. So if `protocolBeforeUpdate` is (3, 7), and `newProtocolForLatestMetadata` is (2, 5), then it will actually add the implicitly supported features.

Reply:
Moved legacy feature normalization in `protocol.merge`. This results into normalizing the protocol after operations such as `ALTER TABLE`, `CLONE`, `RTAS` etc.